src/Sentry.Tunnel/SentryTunnelMiddleware.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Sentry.Tunnel
+{
+    public class SentryTunnelMiddleware : IMiddleware
+    {
+        private readonly string[] _allowedHosts;
+
+        public SentryTunnelMiddleware(string[] allowedHosts)
+        {
+            _allowedHosts = new[] {"sentry.io"}.Concat(allowedHosts).ToArray();
+        }
+
+        public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+        {
+            if (context.Request.Method == "OPTIONS")
+            {
+                context.Response.Headers.Add("Access-Control-Allow-Origin", new[] {(string) context.Request.Headers["Origin"]});
+                context.Response.Headers.Add("Access-Control-Allow-Headers", new[] {"Origin, X-Requested-With, Content-Type, Accept"});
+                context.Response.Headers.Add("Access-Control-Allow-Methods", new[] {"POST, OPTIONS"});
+                context.Response.Headers.Add("Access-Control-Allow-Credentials", new[] {"true"});
+                context.Response.StatusCode = 200;
+                return;
+            }
+
+            var httpClientFactory = context.RequestServices.GetRequiredService<IHttpClientFactory>();
+            var client = httpClientFactory.CreateClient("SentryTunnel");
+            var ms = new MemoryStream();
+            await context.Request.Body.CopyToAsync(ms);
+            ms.Position = 0;
+            using (var reader = new StreamReader(ms))
+            {
+                var header = await reader.ReadLineAsync();
+                if (string.IsNullOrWhiteSpace(header))
+                {
+                    context.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    return;
+                }
+
+                try
+                {
+                    var headerJson = JsonSerializer.Deserialize<Dictionary<string, object>>(header);
+                    if (headerJson == null)
+                    {
+                        context.Response.StatusCode = StatusCodes.Status400BadRequest;
+                        await context.Response.WriteAsync("Invalid DSN JSON supplied");
+                        return;
+                    }
+                    if (headerJson.TryGetValue("dsn", out var dsnString) && Uri.TryCreate(dsnString.ToString(), UriKind.Absolute, out var dsn) && _allowedHosts.Contains(dsn.Host))
+                    {
+                        var projectId = dsn.AbsolutePath.Trim('/');
+                        ms.Position = 0;
+                        var responseMessage = await client.PostAsync($"https://{dsn.Host}/api/{projectId}/envelope/",
+                            new StreamContent(ms));
+                        context.Response.Headers["content-type"] = "application/json";
```

Comment:
You could do the same using the Dsn parser:

```suggestion
                    if (Dsn.TryParse(dsnString) is { } dsn && _allowedHosts.Contains(dsn.ApiBaseUri.Host))
                    {
                        ms.Position = 0;
                        var responseMessage = await client.PostAsync(dsn.GetEnvelopeEndpointUri(),
                            new StreamContent(ms));
                        context.Response.Headers["content-type"] = "application/json";
```

In addition, you need to add the project to the AssemblyInfo from the Sentry project.

Reply:
I guess you can do that, but then I'd have to add Sentry as a dependency and I liked not having that...

Reply:
> In addition, you need to add the project to the AssemblyInfo from the Sentry project.


I don't follow. We're just proxying an event through. 

We can add a User-Agent here btw `Sentry.NET/Tunnel {AssemblyInfo.Version}` or whatever..

Reply:
@bruno-garcia There, added a Sentry.Net Tunnel/{AssemblyInfo.Version} UA